### ISSUE TITLE
Do better filename checking for tleap compatibility

### ISSUE THREAD

**Issue by davidlmobley** [MobleyLab/SolvationToolkit#31]
This is partly an issue with `openmoltools` so I'll raise an issue there as well, but we need to do a better job checking file names for `tleap` compatibility (which is part of the pipeline since we pass through `tleap` to get initial AMBER parameters using tools in `openmoltools.amber`) and raising exceptions if the user provides a filename which won't work. Spaces in file names are not handled properly by `tleap`, but there are also other characters which are problematic. I don't have a full list at this point but I think it includes:
* spaces
* commas
* square brackets?
* french braces?

The most likely reason these characters would creep in is if a user tries to specify a file name which is based on a molecule name. 

@atvanwart - do you have a full list of problematic characters? Or at least a list of all problematic characters you've encountered?

**Comment by atvanwart** [comment]
Yeah, I think I solved this problem for specific cases.  A possible solution is to create a key method and inverse key method.  Amber documentation mentions the characters it does not like to use.  My example below solves the problem for " ()  (  )  )(  ,    "character types.  This is tested against my set of 60+ compound names.  The trick is the order in which replacement takes place as I use just combinations of _ and + for one to one mapping.  

#test for [ old_key== key_map_new(key_map_old(old_key)) ]

#Create key mapping from IUPAC to AMBER(TLEAP)
def key_map_old(old_key):
        A=old_key.replace('+__+','()')
        A=A.replace('_++_',')(')
        A=A.replace('_+',')')
        A=A.replace('+_','(')
        A=A.replace('__',' ')
        A=A.replace('_',',')
        return(A)

#Create key mapping from AMBER(TLEAP) to IUPAC
def key_map_new(new_key):
        B=new_key.replace('()','+__+')
        B=B.replace(')(','_++_')
        B=B.replace(')','_+')
        B=B.replace('(','+_')
        B=B.replace(' ','__')
        B=B.replace(',','_')
        return(B)

**Comment by davidlmobley** [comment]
I was just after a list of which characters cause problems.

For what it's worth, "+" is a valid character in an IUPAC name for some
chiral compounds.

On Thu, Jun 25, 2015 at 11:27 AM, atvanwart <notifications@github.com>
wrote:

> Yeah, I think I solved this problem for specific cases. A possible
> solution is to create a key method and inverse key method. Amber
> documentation mentions the characters it does not like to use. My example
> below solves the problem for " () ( ) )( , "character types. This is tested
> against my set of 60+ compound names. The trick is the order in which
> replacement takes place as I use just combinations of _ and + for one to
> one mapping.
>
> #test for [ old_key== key_map_new(key_map_old(old_key)) ]
>
> #Create key mapping from IUPAC to AMBER(TLEAP)
> def key_map_old(old_key):
> A=old_key.replace('+
>
>
>
> *+','()') A=A.replace('++',')(') A=A.replace('+',')') A=A.replace('+','(')
> A=A.replace('*',' ')
> A=A.replace('_',',')
> return(A)
>
> #Create key mapping from AMBER(TLEAP) to IUPAC
> def key_map_new(new_key):
> B=new_key.replace('()','+
>
>
>
> *+') B=B.replace(')(','++') B=B.replace(')','+') B=B.replace('(','+')
> B=B.replace(' ','*')
> B=B.replace(',','_')
> return(B)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MobleyLab/SolvationToolkit/issues/31#issuecomment-115355022>
> .
>



-- 
David Mobley
dmobley@gmail.com
949-385-2436

**Comment by davidlmobley** [comment]
@atvanwart - just wanted to make sure you saw the above comment, as I forgot to tag you.

Also, for the record (not for you) - I think revisions will be unnecessary here. John Chodera pointed out elsewhere that we can resolve this issue by making tleap work with temporary files and just copy back to the user-specified filenames after done with tleap/antechamber. I went ahead and implemented this here: https://github.com/choderalab/openmoltools/pull/138. So currently anything which is a valid filename ought now to be able to be handled at least on the AMBER side. Probably a few more changes are still needed to make this work all the way through into GROMACS. But soon, I think anything which is a valid filename should be handled properly by the pipeline.

**Comment by atvanwart** [comment]
David,

That is smart way to go (handling it on Amber side).

-Adam

**Comment by davidlmobley** [comment]
Still for me to do here: Make sure that no filenames cause problems for GROMACS/Linux. Probably should try obvious special characters from obscure molecule names.
